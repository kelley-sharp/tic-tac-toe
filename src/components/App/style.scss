.app {
  font-family: 'Martel', serif;
  display: flex;
  flex-flow: column;
  align-items: center;
  margin-top: 3em;
}

.header {
  background-color: #f3f3f3;
  width: 100vw;
  height: 100px;
  display: flex;
  flex-flow: row nowrap;
  justify-content: center;
  align-items: center;
  font-size: 40px;
}

.board {
  margin: 2em;
  display: flex;
  flex-flow: row nowrap;
}

button {
  border-radius: 10px;
  background-color: #a7eee5;
  border-color: #a7eee5;
  color: #05000e;
  width: 100px;
  height: 36px;
  font-family: 'Martel', serif;
  font-size: 15px;
}

button:focus {
  outline: 0;
}

.pending-game-button {
  margin-top: 20px;
}

.board-button-container {
  display: flex;
  flex-direction: column;
  align-items: center;
  margin-top: 1em;
}
