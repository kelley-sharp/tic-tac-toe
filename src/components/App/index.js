import React, { Component } from 'react';
import './style.scss';
import Board from '../Board';
import didWin from '../../helpers/didWin';
import processTurn from '../../helpers/processTurn';

const getDefaultState = () => ({
  gamePending: true,
  gameOver: false,
  currentPlayer: 'X',
  turn: 1,
  message: '',
  buttonText: 'Start game',
  board: [[null, null, null], [null, null, null], [null, null, null]],
  winner: null
});

class App extends Component {
  constructor(props) {
    super(props);
    this.state = getDefaultState();
  }

  handleTurn = (row, col) => {
    let { updatedBoard, nextPlayer, nextMessage, nextTurn } = processTurn(
      this.state.board,
      { row, col },
      this.state.currentPlayer,
      this.state.turn
    );
    //if the updated board shows someone winning, save that info in related variables
    let winner = null;
    let gameOver = false;
    let nextButtonText = 'End game';

    const hasWinner = didWin(updatedBoard);
    if (hasWinner) {
      winner = this.state.currentPlayer;
      gameOver = true;
      nextMessage = `${this.state.currentPlayer} wins!`;
      nextButtonText = 'Play again';
    } else if (!hasWinner && nextTurn === 10) {
      gameOver = true;
      nextMessage = "It's a scratch, no one wins.";
      nextButtonText = 'Play again';
    }

    //Update the state with new information
    this.setState({
      board: updatedBoard,
      currentPlayer: nextPlayer,
      message: nextMessage,
      turn: nextTurn,
      winner: winner,
      gameOver: gameOver,
      buttonText: nextButtonText,
      frozen: gameOver
    });
  };

  handleGameBegin = () => {
    this.setState({
      gamePending: false,
      message: 'Player one (X) - make the first mark!',
      buttonText: 'End game'
    });
  };

  resetGame = () => {
    this.setState(getDefaultState);
  };

  render() {
    return (
      <div className="app">
        {this.state.gamePending ? (
          <div>
            <button
              className="pending-game-button"
              onClick={this.handleGameBegin}
            >
              {this.state.buttonText}
            </button>
            <div className="message">
              <p>{this.state.message}</p>
            </div>
            <div className="board-placeholder" />
          </div>
        ) : (
          <div className="board-button-container">
            <button onClick={this.resetGame}>{this.state.buttonText}</button>
            <div className="message">
              <p>{this.state.message}</p>
            </div>
            <Board
              board={this.state.board}
              frozen={this.state.gameOver}
              handleTurn={this.handleTurn}
            />
          </div>
        )}
      </div>
    );
  }
}

export default App;
