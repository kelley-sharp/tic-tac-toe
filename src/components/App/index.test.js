import React from 'react';
import App from '.';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

describe('<App />', () => {
  // smoke test
  it('renders without crashing', function() {
    shallow(<App />);
  });

  // snapshot test
  it('matches snapshot', function() {
    let wrapper = shallow(<App />);
    let serialized = toJson(wrapper);
    expect(serialized).toMatchSnapshot();
  });
});
