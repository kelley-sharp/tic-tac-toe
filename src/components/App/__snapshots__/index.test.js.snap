// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> matches snapshot 1`] = `
<div
  className="app"
>
  <div>
    <button
      className="pending-game-button"
      onClick={[Function]}
    >
      Start game
    </button>
    <div
      className="message"
    >
      <p />
    </div>
    <div
      className="board-placeholder"
    />
  </div>
</div>
`;
